### DISCUSSION TITLE
Create link to another page/place in pdf

### DISCUSSION THREAD

**Discussion by chrisjd20** [pymupdf/PyMuPDF#2065] [Q&A] [answered]
The documentation is very sparse on how it would be possible to create a link to another page/location in the same pdf.

For example, I need to insert some text into a specific location on the page with embedded link to another page in the same pdf:

```
Please click [here] for more information.
```

Where `[here]` is the link. `here` being the link text. This link would redirect the user to page 10.

**Comment by JorjMcKie** [comment]
So what would you like to see?
An example code snippet?

**Comment by JorjMcKie** [comment]
So what would you like to see?
An example code snippet?

**Comment by JorjMcKie** [comment, marked as answer]
Here is some help for your example:
1. Locate the area you wish to cover with the link. FYI: this does not need to be text at all, could be an area covering an image or even empty space. If you want text, you must locate its bbox.
2. Define a dictionary `{"kind": fitz.LINK_GOTO, "page": 9, "from": area, "to": point}`. Where "point" is a `fitz.Point` on the target page.
3. Execute `page.insert_link(... this dictionary ...)`.

We will include this "recipe" in the documentation.

**Reply by chrisjd20** [reply]
Ok, thanks for the help. To clarify, let's say I wanted to have a 10x10 area/text (with the target page num) at x=200 and y=200  that if someone clicked on it would take them to page 9 at point `Point(0.0, 191.0)`.

For example:
``` python
PDFHEIGHT = 792
PDFWIDTH = 612
a = 0
b = 0
c = 0
d = 0
rect = fitz.Rect(a, b, c, d)
page.insert_link(
  {"kind": fitz.LINK_GOTO, "page": 9, "from": rect, "to": fitz.Point(0.0, 191.0)}
)
# what is  a,b,c,d given I need a 10x10 area located at x=200 and y=200 on the page.
```

How would I calculate a,b,c,d given my area/location requirements mentioned in my first sentence (above) ?

In fact, maybe there is a function that can generate a `Rect` given the height/width and x/y location requirements?

**Reply by JorjMcKie** [reply]
Rectangles in PyMuPDF are given by either 4 float coordinates, or 2 points, or 1 point and 2 floats.
In any case, this leads to 4 float coordinates for (1) the north-western point, and (2) the south-eastern point.

So in your example `rect = fitz.Rect(200, 200, 210, 210)`.
Or, equivalently `fitz.Rect(a, b, a + width, b + height)`.
And because you can use "geometry" objects as operands in algebraic expressions, the following is also possible:

``rect = fitz.Rect(a, b, a, b) + (0, 0, width, height)``. Here the second operand may be any sequence of 4 floats, to be added component-wise.

You get the point?

**Reply by chrisjd20** [reply]
OHHHHHHHHHHHHHH ok that's so simple. I was way over thinking it. Thanks!

**Reply by JorjMcKie** [reply]
If you first need to locate the text that should become the "hot area" of a link, first search for it, then use the returned hit rectangles:
```python
rl = page.search_for("[here]")
for rect in rl:
    # make a link for this hit
    # etc.
```

**Reply by chrisjd20** [reply]
Thanks that's a nice addition to my question! :)

**Comment by julian-smith-artifex-com** [comment]
This example is documented in the latest release, PyMuPDF-1.21.1, e.g. see: https://pymupdf.readthedocs.io/en/latest/page.html#writing
